core/src/com/unciv/ui/worldscreen/unit/UnitActions.kt

```diff
@@ -299,96 +303,86 @@ object UnitActions {
-                 * The only known cause of this currently is "land units upgrading to water units" which fail to be placed.
-                 */
-                if (newUnit == null) {
-                    val readdedUnit = unit.civInfo.placeUnitNearTile(unitTile.position, unit.name)
-                    unit.copyStatisticsTo(readdedUnit!!)
-                } else { // Managed to upgrade
-                    unit.civInfo.addGold(-goldCostOfUpgrade)
-                    unit.copyStatisticsTo(newUnit)
-                    newUnit.currentMovement = 0f
-                }
-            }.takeIf {
-                unit.civInfo.gold >= goldCostOfUpgrade
-                && unit.currentMovement > 0
-                && !unit.isEmbarked()
-            }
-        )
-    }
-    
-    fun getFreeUpgradeAction(unit: MapUnit): UnitAction? {
-        if (unit.baseUnit().upgradesTo == null) return null
-        val upgradedUnit = unit.civInfo.getEquivalentUnit(unit.baseUnit().upgradesTo!!)
-        if (!unit.canUpgrade(upgradedUnit, true)) return null
+        val goldCostOfUpgrade = if (isFree) 0 else unit.getCostOfUpgrade(upgradedUnit)
+
+        // No string for "FREE" variants, these are never shown to the user.
+        // The free actions are only triggered via OneTimeUnitUpgrade or OneTimeUnitSpecialUpgrade in UniqueTriggerActivation.
+        val title = if (deltaStr.isEmpty()) "Upgrade to [${upgradedUnit.name}] ([$goldCostOfUpgrade] gold)"
+            else "Upgrade to [${upgradedUnit.name}]\n([$goldCostOfUpgrade] gold, [$deltaStr])"
 
         return UnitAction(UnitActionType.Upgrade,
-            title = "Upgrade to [${upgradedUnit.name}] (FREE)",
+            title = title,
             action = {
-                val unitTile = unit.getTile()
                 unit.destroy()
-                val newUnit = unit.civInfo.placeUnitNearTile(unitTile.position, upgradedUnit.name)
+                val newUnit = civInfo.placeUnitNearTile(unitTile.position, upgradedUnit.name)
 
                 /** We were UNABLE to place the new unit, which means that the unit failed to upgrade!
                  * The only known cause of this currently is "land units upgrading to water units" which fail to be placed.
                  */
                 if (newUnit == null) {
-                    val readdedUnit = unit.civInfo.placeUnitNearTile(unitTile.position, unit.name)
-                    unit.copyStatisticsTo(readdedUnit!!)
+                    val resurrectedUnit = civInfo.placeUnitNearTile(unitTile.position, unit.name)!!
+                    unit.copyStatisticsTo(resurrectedUnit)
                 } else { // Managed to upgrade
+                    if (!isFree) civInfo.addGold(-goldCostOfUpgrade)
                     unit.copyStatisticsTo(newUnit)
                     newUnit.currentMovement = 0f
                 }
+            }.takeIf {
+                isFree || (
```

Comment:
This looks like an error, and should be (isFree || unit.civInfo.gold >= goldCostOfUpgrade)
If a unit is free it shouldn't allow me to upgrade when embarked or if I can't upgrade

Reply:
Then it was an error in the 'before' state too. isFree is only ever used in mods or for ruins at the moment, and ruins guarantee not embarked while it's also clear the unit should still upgrade when it has just expended all movement points... Ruins also definitely must bypass the normal canUpgrade check, and the other free unique.. Yes it must also bypass that (it's there for resource cost) prefer going into resource debt instead of silently wasting a free upgrade? I'd say yes. We'll get there when modders complain.

Reply:
Fair enough